contribute/design/1.x/cstor/20200511-cstor-backupandrestore.md

```diff
@@ -0,0 +1,259 @@
+2. Execute the REST API(POST `/latest/backups`) of CVC-Operator (This object will include the IP Address of snapshot receiver/sender module) to create backup.
+    2.1 Create a snapshot using volume name and snapshot name(which will get during the CreateSnapshot request).
+    2.2 Find the Healthy cStorVolumeReplica if it doesn't exist return error.
+    2.3 Create CStorCompleteBackUp resources if it doesn’t exist(Intention of creating this resource is used for incremental backup purpose).
+    2.4 Create CStorBackUp resource by populating the current snapshot name and previous snapshot name(if exists from CStorCompleteBackUp) with Healthy CStorVolumeReplica pool UID.
+    2.5 Corresponding backup controllers exist in pool-manager will responsible for sending the snapshot data from pool to velero-plugin.         and velero-plugin will write this stream to cloud-provider.
+3. Call cloud interface API `UploadSnapshot` which will upload the backup to the cloud provider.
+4. This API will return the unique ‘snapshotID’ (volumeID + "-velero-bkp-" + backupName) to the velero server. This ‘snapshotID’ will be used to refer to the backup snapshot.
+
+##### Velero Server Sends DeleteSnapshot Request
+
+Velero server sends `DeleteSnapshot` API of Velero Interface to delete the backup/snapshot with argument `snapshotID`. This snapshotID is generated during the backup creation of this snapshot. Velero Interface will execute the DeleteSnapshot API of cStor backup/restore module. cStor backup/restore module is responsible for performing below steps:
+1. Delete the PVC object for this backup from the cloud provider.
+2. Execute REST API(`latest/backups/`) of the CVC-Operater to delete the resources created for this particular backup.
+    2.1 Delete the CStorCompleteBackup resources if it is not a complete backup or if that is the only backup exists for that volume(So next backup will be complete backup).
+    2.2 Delete the snapshot created for that backup.
+    2.3 Delete the CStorBackUp resource.
+3. Execute the `RemoveSnapshot` API of the cloud interface to delete the snapshot from the cloud provider.
+
+##### Velero Server Sends CreateVolumeFromSnapshot Request
+
+Velero Server will execute the `CreateVolumeFromSnapshot` API of the velero interface to restore the backup with the argument (snapshotID, volumeType). Velero interface will execute `CreateVolumeFromSnapshot` API of cStor backup/restore module. cStor backup/restore module will perform following below steps:
+1. Download the PVC object from the cloud provider through a cloud interface and deploy the PVC. If PVC already exists in the cluster then skip the PVC creation(only for remote restore).
+2. Check If PVC status is bounded.
+3. Execute the REST API(`latest/restore`) with CVC-Operator restore details(includes the IP address of snapshot receiver/sender module) to initiate the restore of the snapshot.
+   3.1 Create CVC if it doesn't exist in the cluster.
```

Comment:
cvc should have annotation, which can be passed to cvr, to instruct cvr controller about skipping target-ip address assignement to cvr.

Reply:
@mynktl can you confirm below points:
- If it is remote restore then PVC will be created by velero-plugin then corresponding volume provisioner will create PV and CVC(with annotation) then all the cstorvolume resources will be created by CVC-controller. Then restore endpoint will be executed by velero-plugin.
-  If it is a local restore request then first we are creating CVC with `restoreObj.Spec.RestoreName`(snapshot name) in this case we don't need to set annotation `openebs.io/created-through`.

Reply:
> @mynktl can you confirm below points:
> 
>     * If it is remote restore then PVC will be created by velero-plugin then corresponding volume provisioner will create PV and CVC(with annotation) then all the cstorvolume resources will be created(especially if the annotation is set then CVR will have isRestoreVol annotation) by CVC-controller. Then restore endpoint will be executed by velero-plugin.
>
:+1:  
>     * If it is a local restore request then first we are creating CVC with `restoreObj.Spec.RestoreName`(snapshot name) in this case we don't need to set annotation `openebs.io/created-through`.
:+1: